index/segment/fst/reader.go

```diff
@@ -0,0 +1,354 @@
+	}
+
+	fieldsFST, err := vellum.Load(opts.FSTFieldsData)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load fields fst: %v", err)
+	}
+	return &reader{
+		fieldsFST: fieldsFST,
+
+		opts:    opts,
+		numDocs: metadata.NumDocs,
+	}, nil
+}
+
+type reader struct {
+	sync.RWMutex
+	closed bool
+
+	fieldsFST *vellum.FST
+
+	opts    NewReaderOptions
+	numDocs int64
+}
+
+func (r *reader) Size() int64 {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return 0
+	}
+	return r.numDocs
+}
+
+func (r *reader) ContainsID(docID []byte) (bool, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return false, errReaderClosed
+	}
+	pl, err := r.MatchTerm(doc.IDReservedFieldName, docID)
+	if err != nil {
+		return false, err
+	}
+	return !pl.IsEmpty(), nil
+}
+
+func (r *reader) Reader() (index.Reader, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+	return r, nil
+}
+
+func (r *reader) Close() error {
+	r.Lock()
+	defer r.Unlock()
+	if r.closed {
+		return errReaderClosed
+	}
+	r.closed = true
+	return r.fieldsFST.Close()
+}
+
+// FOLLOWUP(prateek): really need to change the types returned in Fields() and Terms()
+// to be iterators to allow us to pool the bytes being returned to the user. Otherwise
+// we're forced to copy these massive slices every time.
+func (r *reader) Fields() ([][]byte, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	return r.allKeys(r.fieldsFST)
+}
+
+func (r *reader) Terms(field []byte) ([][]byte, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	termsFST, err := r.retrieveTermsFST(field)
+	if err != nil {
+		return nil, err
+	}
+
+	terms, err := r.allKeys(termsFST)
+	if err != nil {
+		termsFST.Close()
+		return nil, err
+	}
+
+	if err := termsFST.Close(); err != nil {
+		return nil, err
+	}
+
+	return terms, nil
+}
+
+func (r *reader) MatchTerm(field []byte, term []byte) (postings.List, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	termsFST, err := r.retrieveTermsFST(field)
+	if err != nil {
+		return nil, err
+	}
+	defer termsFST.Close()
+
+	postingsOffset, exists, err := termsFST.Get(term)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		// i.e. we don't know anything about the term, so can early return an empty postings list
+		return r.opts.PostingsListPool.Get(), nil
+	}
+
+	return r.retrievePostingsList(postingsOffset)
+}
+
+func (r *reader) MatchRegexp(field []byte, regexp []byte, compiled *regexp.Regexp) (postings.List, error) {
+	re, err := vregex.New(string(regexp))
+	if err != nil {
+		return nil, err
+	}
+
+	termsFST, err := r.retrieveTermsFST(field)
+	if err != nil {
+		return nil, err
+	}
+	defer termsFST.Close()
+
+	pl := r.opts.PostingsListPool.Get()
+
+	iter, iterErr := termsFST.Search(re, minByteKey, maxByteKey)
+	defer iter.Close()
```

Comment:
nit: Combine the defers?

Reply:
Also do you need to check return type of `iter.Close()` here?

Reply:
sure thing, started using the SafeCloser idiom here

Reply:
done